lib/consumer.js

```diff
@@ -5,82 +7,77 @@ var amqp = require('amqplib'),
+          amqpConnection = null;
+          amqpChannel = null;
+        });
 
-        amqpConnection.on('close', reconnect);
-        amqpConnection.on('error', reconnect);
+        _connection.on('error', Logger.error);
 
-        return amqpConnection.createChannel();
+        amqpConnection = _connection;
+        resolve(amqpConnection);
       })
-      .then(function (_channel) {
-        amqpChannel = _channel;
-        amqpChannel.prefetch(amqpConfig.amqpPrefetch);
+      .catch((err) => {
+        amqpConnection = null;
+        reject(err);
+      });
+    });
+  }
+
+  return amqpConnection;
+};
 
-        Logger.info('[BMQ-CONSUMER] Is now connected and ready to consume messages');
+var createChannel = () => {
+  if (!amqpChannel) {
+    amqpChannel = new Promise((resolve, reject) => {
+      amqpConnection.createChannel()
+        .then((_channel) => {
+          _channel.prefetch(amqpConfig.amqpPrefetch);
 
-        amqpState = amqpStates.CONNECTED;
+          Logger.info('[BMQ-CONSUMER] Is now connected and ready to consume messages');
 
-        amqpIntervalId = clearInterval(amqpIntervalId);
+          _channel.on('close', (err) => {
+            if (err) Logger.error('ERROR:', err);
+            amqpChannel = null;
+          });
 
-        amqpChannel.on('close', reconnect);
-        amqpChannel.on('error', reconnect);
+          _channel.on('error', Logger.error);
 
-        while (amqpConnPool.length > 0) {
-          amqpConnPool.shift()(amqpChannel);
-        }
+          amqpChannel = _channel;
+          resolve(amqpChannel);
 
-        if (amqpReconnection) {
-          amqpReconnection = false;
-          for (var i = 0, l = amqpConsumptions.length; i < l; ++i) {
-            consume(amqpConsumptions[i].queue, amqpConsumptions[i].options, amqpConsumptions[i].callback);
+          if (amqpReconnect) {
+            amqpReconnect = false;
+            for (var i = 0, l = amqpQueues.length; i < l; ++i) {
+              consume(amqpQueues[i].queue, amqpQueues[i].options, amqpQueues[i].callback);
+            }
           }
-        }
+        })
+        .catch((err) => {
+          amqpChannel = null;
+          reject(err);
+        });
+    });
+  }
 
-        _resolve(amqpChannel);
-      })
-      .catch(reconnect);
-    } else if (amqpState === amqpStates.CONNECTING) {
-      amqpConnPool.push(_resolve);
-    } else {
-      _resolve(amqpChannel);
-    }
-  });
-};
-var reconnect = function(err) {
-  if (err) Logger.error('[BMQ-CONSUMER]', err);
-
-  amqpIntervalId = (!amqpIntervalId) ? clearInterval(amqpIntervalId) : amqpIntervalId;
-  Logger.info('[BMQ-CONSUMER] Try to reconnect');
-  amqpState = amqpStates.DISCONNECTED;
-  amqpReconnection = true;
-  amqpIntervalId = setInterval(connect, 1000);
+  return amqpChannel;
 };
 
-var checkRpc = function(msg, queue) {
+var checkRpc = (msg, queue) => {
   return function (_content) {
     if (_content !== undefined && msg.properties.replyTo) {
```

Comment:
Because we stated that null could be a returnable value.

In an RPC call, we could see something like this:

```js
produce('getUser', { userId: 0}, { rpc: true } )
  .then((user) => {
    //user can be null if not found, but not undefined
  });
```

Because we stated that a service should not send back undefined because this means it should be RPC, or the service does not makes sense, because undefined means nothing. We based this on our usage, maybe we are wrong.

It should be documented too!